spyder_terminal/widgets/main_widget.py

```diff
@@ -0,0 +1,451 @@
+
+    # --- Signals
+    # ------------------------------------------------------------------------
+    sig_server_is_ready = Signal()
+    """
+    This signal is emitted when the server is ready to connect.
+    """
+
+    def __init__(self, name, plugin, parent):
+        """Widget constructor."""
+        self.terms = []
+        super().__init__(name, plugin, parent)
+
+        # Attributes
+        self.tab_widget = None
+        self.menu_actions = None
+        self.server_retries = 0
+        self.server_ready = False
+        self.font = None
+        self.port = select_port(default_port=8071)
+        self.server_stdout = subprocess.PIPE
+        self.server_stderr = subprocess.PIPE
+        self.stdout_file = osp.join(os.getcwd(), 'spyder_terminal_out.log')
+        self.stderr_file = osp.join(os.getcwd(), 'spyder_terminal_err.log')
+        if get_debug_level() > 0:
+            self.server_stdout = open(self.stdout_file, 'w')
+            self.server_stderr = open(self.stderr_file, 'w')
+        self.project_path = None
+        self.current_file_path = None
+        self.current_cwd = os.getcwd()
+
+        # Widgets
+        self.main = parent
+        self.find_widget = FindTerminal(self)
+        self.find_widget.hide()
+
+        layout = QVBoxLayout()
+
+        # Tab Widget
+        self.tabwidget = Tabs(self, rename_tabs=True)
+        self.tabwidget.currentChanged.connect(self.refresh_plugin)
+        self.tabwidget.move_data.connect(self.move_tab)
+        self.tabwidget.set_close_function(self.close_term)
+
+        if (hasattr(self.tabwidget, 'setDocumentMode') and
+                not sys.platform == 'darwin'):
+            # Don't set document mode to true on OSX because it generates
+            # a crash when the console is detached from the main window
+            # Fixes Issue 561
+            self.tabwidget.setDocumentMode(True)
+        layout.addWidget(self.tabwidget)
+        layout.addWidget(self.find_widget)
+        self.setLayout(layout)
+
+        self.__wait_server_to_start()
+
+    # ---- PluginMainWidget API
+    # ------------------------------------------------------------------------
+    def get_focus_widget(self):
+        """
+        Set focus on current selected terminal.
+
+        Return the widget to give focus to when
+        this plugin's dockwidget is raised on top-level.
+        """
+        term = self.tabwidget.currentWidget()
+        if term is not None:
+            return term.view
+
+    def get_title(self):
+        """Define the title of the widget."""
+        return _('Terminal')
+
+    def setup(self):
+        """Perform the setup of plugin's main menu and signals."""
+        self.cmd = find_program(self.get_conf('shell'))
+        self.server = subprocess.Popen(
+            [sys.executable, '-m', 'spyder_terminal.server',
+             '--port', str(self.port), '--shell', self.cmd],
+            stdout=self.server_stdout,
+            stderr=self.server_stderr)
```

Comment:
Perhaps we should use QProcess instead of subprocess to be able to tell if the server started correctly? This is just something to consider for later, not for this PR.

Reply:
I will open a new issue with this so we don't forget